lib/Dialect/TritonGPU/IR/Ops.cpp

```diff
@@ -43,26 +53,48 @@ struct CanonicalizeConvertFromReshape
     if (!convert)
       return failure();
     // If the layouts are structurally the same, the convert is trivial
-    auto srcType = convert.getSrc().getType();
-    auto dstType = convert.getType();
-    auto srcLL = toLinearLayout(srcType.getShape(), srcType.getEncoding());
-    auto dstLL = toLinearLayout(dstType.getShape(), dstType.getEncoding());
-    if (srcLL == dstLL) {
+    if (isConvertTrivial(convert)) {
       rewriter.replaceOpWithNewOp<triton::ReshapeOp>(
-          op, op.getType(), convert.getSrc(), op.getAllowReorder());
-      return mlir::success();
+          op, op.getType(), convert.getSrc(), op.getAllowReorder(),
+          op.getEfficientLayout());
+      return success();
     }
+
     if (isExpensiveView(convert.getSrc().getType(), op.getType()))
       return failure();
     if (!op.getAllowReorder() || op.getEfficientLayout())
       return failure();
 
     rewriter.replaceOpWithNewOp<triton::ReshapeOp>(
-        op, op.getType(), convert.getSrc(), op.getAllowReorder());
+        op, op.getType(), convert.getSrc(), op.getAllowReorder(),
+        op.getEfficientLayout());
     return mlir::success();
   }
 };
 
+// TODO We should do this generically for op(cvt) -> op
+// We have similar patterns for reshape and split...
+// See https://github.com/triton-lang/triton/pull/5403#discussion_r1920091671
+
+// trans(cvt) -> trans
+struct CanonicalizeConvertFromTranspose
+    : public mlir::OpRewritePattern<triton::TransOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(triton::TransOp op,
+                  PatternRewriter &rewriter) const override {
+    // If the layouts are structurally the same, the convert is trivial
+    auto convert = op.getSrc().getDefiningOp<ConvertLayoutOp>();
+    if (!convert || !isConvertTrivial(convert))
+      return failure();
```

Comment:
thinking more about this having a canonical form for layouts would be the ideal solution and solve this without extra patterns. I think this is a similar idea of what you mentioned about being able to convert linear to legacy layout.

I'm a bit worried the equivalent layout checks are going to spread around the code and make things more complicated than needed.

Anyway, nothing to do in this PR for it.

Reply:
If we can overload `operator==` an `Attribute` that should be easy to do. Once we make DistributedEncoding inherit from LinearEncoding, we would implement equality of encodings by doing the check on the LLs, and then implement equality of Types by implementing equality via `toLinearLayout` (this is a bit more generic than the check above but it's not 100% necessary if it's to big a hassle).

I just wonder whether MLIR allows us to do this...

Reply:
I don't think we can/want to override operator== for an attribute type. What I was hoping for is a way that for a given shape and layout picks a canonical form picks a canonical form of any layouts. That would avoid having to handle multiple layouts that represent the same thing.
Anyway we can chat more about it at some point

Reply:
Yes, we can totally do that, in fact, that's what `toLinearLayout` does ATM.
The issue with this is that the layout would be then associated to the RankedTensorType. If we do that, it would be tricky to keep the "tiling" semantics that we have when we broadcast a layout. In particular, BroadcastOp could not be `SameOperandsAndResultEncoding` any longer. That being said, I agree this would be a more natural representation of layouts.

But yeah, we can discuss at a later point.